framework/src/modules/interoperability/mainchain/module.ts

```diff
@@ -227,4 +244,191 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
+			genesisInteroperabilitySchema,
+			genesisBlockAssetBytes,
+		);
+
+		validator.validate<GenesisInteroperability>(
+			genesisInteroperabilitySchema,
+			genesisInteroperability,
+		);
+
+		const { ownChainName, ownChainNonce, chainInfos } = genesisInteroperability;
+
+		// On the mainchain, the following checks are performed:
+		if (ctx.chainID.equals(getMainchainID(ctx.chainID))) {
+			// ownChainName == CHAIN_NAME_MAINCHAIN.
+			if (ownChainName !== CHAIN_NAME_MAINCHAIN) {
+				throw new Error(`ownChainName must be equal to CHAIN_NAME_MAINCHAIN'`);
+			}
+
+			// if chainInfos is empty, then ownChainNonce == 0
+			if (chainInfos.length === 0) {
+				if (ownChainNonce !== BigInt(0)) {
+					throw new Error(`ownChainNonce must be 0 if chainInfos is empty.`);
+				}
+			} else {
+				// If chainInfos is non-empty
+				// ownChainNonce > 0
+				if (ownChainNonce <= 0) {
+					throw new Error(`ownChainNonce must be positive if chainInfos is not empty.`);
+				}
+
+				// Each entry chainInfo in chainInfos has a unique chainInfo.chainID
+				const chainIDs = chainInfos.map(info => info.chainID);
+				if (!bufferArrayUniqueItems(chainIDs)) {
+					throw new Error(`chainInfos doesn't hold unique chainID.`);
+				}
+
+				// chainInfos is ordered lexicographically by chainInfo.chainID
+				const sortedByChainID = [...chainInfos].sort((a, b) => a.chainID.compare(b.chainID));
+				for (let i = 0; i < chainInfos.length; i += 1) {
+					if (!chainInfos[i].chainID.equals(sortedByChainID[i].chainID)) {
+						throw new Error('chainInfos is not ordered lexicographically by chainID.');
+					}
+				}
+
+				this._verifyChainInfos(ctx, chainInfos);
+			}
+		}
+	}
+
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#mainchain
+	private _verifyChainInfos(ctx: GenesisBlockExecuteContext, chainInfos: ChainInfo[]) {
+		const mainchainID = getMainchainID(ctx.chainID);
+		const chainDataNames = chainInfos.map(info => info.chainData.name);
+
+		// verify root level properties
+		for (const chainInfo of chainInfos) {
+			const { chainID } = chainInfo;
+
+			// chainInfo.chainID != getMainchainID();
+			if (chainID.equals(mainchainID)) {
+				throw new Error(`chainID must be not equal to getMainchainID().`);
+			}
+
+			// - chainInfo.chainId[0] == getMainchainID()[0].
+			if (chainID[0] !== mainchainID[0]) {
+				throw new Error(`chainID[0] doesn't match getMainchainID()[0].`);
+			}
+
+			this._verifyChainData(ctx, chainInfo, chainDataNames);
+			this._verifyChannelData(ctx, chainInfo);
+			this._verifyChainValidators(chainInfo);
+		}
+	}
+
+	private _verifyChainData(
+		ctx: GenesisBlockExecuteContext,
+		chainInfo: ChainInfo,
+		chainDataNames: string[],
+	) {
+		const validStatuses = [ChainStatus.REGISTERED, ChainStatus.ACTIVE, ChainStatus.TERMINATED];
+
+		// The entries chainData.name must be pairwise distinct
+		if (new Set(chainDataNames).size !== chainDataNames.length) {
```

Comment:
If I understand well, this checks if among all items of `chainInfos` array, the `chainData.name` properties are pairwise distinct. If that's the case, I don't see why this check is performed here; since this function is called by `verifyChainInfos` for all items in `chainInfos`, we perform the same check multiple times (if the size of `chainInfos` is `m`, we perform it `m` times). 

I guess this check could be done just once, even before calling `verifyChainInfos` (e.g., in line 300 after checking that `chainInfos` is sorted w.r.t `chainID` and no duplicates exist)

Reply:
Understand, in that case I would say it is safe to remove this checking here

Reply:
I see you removed this check from here, but it is not added in `initGenesisState` (e.g., line 300) so it is like this check is not done at all now..

Reply:
Ah understand, added this check under `initGenesisState`